File: student_Banasthali_Nivedita/graphs/problem10.cpp
/*
10. Implement a function to find the number of connected components in an undirected graph.
*/

#include<bits/stdc++.h>

using namespace std;

void dfs(vector<vector<int>>& graph, int node, vector<bool>& visited) {
    visited[node] = true;

    for (int neighbor : graph[node]) {
        if (!visited[neighbor]) {
            dfs(graph, neighbor, visited);
        }
    }
}

int countConnectedComponents(int n, vector<vector<int>>& edges) {
    vector<vector<int>> graph(n);
    vector<bool> visited(n, false);
    int count = 0;

    //adjacency list representation of the graph
    for (const vector<int>& edge : edges) {
        int u = edge[0];
        int v = edge[1];
        graph[u].push_back(v);
        graph[v].push_back(u);
    }

    // DFS to count the number of connected components
    for (int i = 0; i < n; i++) {
        if (!visited[i]) {
            count++;
            dfs(graph, i, visited);
        }
    }

    return count;
}

int main() {
    int n = 5;
    vector<vector<int>> edges = {
        {0, 1},
        {1, 2},
        {3, 4}
    };

    int numComponents = countConnectedComponents(n, edges);
    cout << "Number of connected components: " << numComponents << endl;

    return 0;
}

File: ShrutiVarshney_BanasthaliVidyapith_GraphsAssignment/GraphsAssignmentQuestions/src/graphsassignmentquestions/Problem10.java
package graphsassignmentquestions;
import java.util.*;
public class Problem10 
{
    public int countComponents(int n, int[][] edges)
    {
        // Create adjacency list from the input edges
        List<Integer>[] adjList = new List[n];
        for (int i = 0; i < n; i++)
        {
            adjList[i] = new ArrayList<>();
        }
        for (int[] edge : edges)
        {
            int u = edge[0];
            int v = edge[1];
            adjList[u].add(v);
            adjList[v].add(u); // For undirected graph, add both directions
        }

        // Create an array to mark visited nodes
        boolean[] visited = new boolean[n];

        // Perform DFS to count the connected components
        int count = 0;
        for (int i = 0; i < n; i++)
        {
            if (!visited[i])
            {
                dfs(i, adjList, visited);
                count++;
            }
        }

        return count;
    }

    private void dfs(int node, List<Integer>[] adjList, boolean[] visited)
    {
        visited[node] = true;
        for (int neighbor : adjList[node]) 
        {
            if (!visited[neighbor])
            {
                dfs(neighbor, adjList, visited);
            }
        }
    }

    public static void main(String[] args)
    {
        int n = 5;
        int[][] edges = {{0, 1}, {1, 2}, {3, 4}};

        Problem10 solution = new Problem10();
        int count = solution.countComponents(n, edges);
        System.out.println(count); // Output: 2
    }
}
